### PR TITLE
Compatibility with Kibana 7.12.1

### PR DESCRIPTION
**Pull request by cpiment**
Issue #197 asks for Kibana 7.12+ support for the plugin. This WIP PR shows the current efforts in doing that possible for Kibana 7.12.1

At the current state there are at least two features that do not work:

- [x] Document Table visualization
- [x] Computed Columns
- [x] Scrollbars are not shown when visualizations in dashboard
- [x] Data tab of the Inspect window is empty
- [x] Filter for / filter out value don't work

### CODE REVIEW

**Comment by cpiment** [comment]
Fixed "Add computed column" :)

**Comment by cpiment** [comment]
Fixed "Document Table"

**Comment by fbaligand** [comment]
After a first test, I realized there is a little missing change:
in enhanced-table-vis-controller.js, `$scope.vis.params` does not exist anymore.
You have to replace it by `$scope.visParams`

**Comment by cpiment** [comment]
While I was testing the package building I realised just the same. I saw in other parts of the code that you use `$scope.visState.params` and I used that. Is better to use `$scope.visParams`?

**Comment by fbaligand** [comment]
Well, to my known, $scope.visState is editor state when user has not "applied changes".
$scope.visParams is "saved state".
So I would prefer "$scope.visParams" for this precise case.

**Comment by fbaligand** [comment]
I looked at `doc_table_fn.ts` and `enh_table_fn.ts`. Globally, there is a bunch of copy/paste code between these 2 files. Too much :)
Actually, this is a backport of previous core `visualization.ts`.
And I think it should be as is. I mean:
- only one file, called `visualization.ts`
- the response handler is an input parameter
- and both enhanced and document response handlers do not need for a "dimensions" input parameter

**Comment by fbaligand** [comment]
For more coherence with existing files, could you:
- rename `doc_table_*` to `document-table-*`
- rename `enh_table_*` to `enhanced-table-*`

File: public/data_load/enhanced-table-request-handler.js [hunk 1]
```diff
@@ -85,8 +84,10 @@ export async function enhancedTableRequestHandler ({
     });
   }
 
-  inspectorAdapters.requests = new RequestAdapter();
-  inspectorAdapters.data = new DataAdapter();
+  //inspectorAdapters.requests is already initialized
+  //inspectorAdapters.requests = new RequestAdapter();
+  // DataAdapter is not present in Kibana 7.12. Is this necessary?
+  //inspectorAdapters.data = new DataAdapter();
```
Comment (fbaligand) [inline public/data_load/enhanced-table-request-handler.js:90]: it's maybe somewhat necessary, because "Inspect" view don't show anymore data into "View: data" :(
You should take inspiration from Kibana 7.12 "esaggs" component.

File: public/data_load/enhanced-table-request-handler.js [hunk 2]
```diff
@@ -113,9 +114,13 @@ export async function enhancedTableRequestHandler ({
...
   response.aggs = aggs;
-  response.columns.forEach(column => {
+  // Columns already have meta in Kibana 7.12
+/*   response.columns.forEach(column => {
     column.meta = serializeAggConfig(column.aggConfig);
-  });
+  }); */
+
+  //Add aggConfig to each column
+  response.columns = await enrichColumnsWithAggconfig(response.columns)
```
Comment (fbaligand) [inline public/data_load/enhanced-table-request-handler.js:123]: instead of creating a new AggConfig object, you should get existing one in `response.aggs` object, like this:

``` javascript
  response.columns.forEach(column => {
    column.aggConfig = column.aggs.byId(column.meta.sourceParams.id);
  });
```

File: public/enh_table_to_ast.ts [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+
+  if (visParams.showPartialRows && !visParams.showMetricsAtAllLevels) {
+    // Handle case where user wants to see partial rows but not metrics at all levels.
+    // This requires calculating how many metrics will come back in the tabified response,
+    // and removing all metrics from the dimensions except the last set.
+    const metricsPerBucket = metrics.length / buckets.length;
+    visConfig.dimensions.metrics.splice(0, metricsPerBucket * buckets.length - metricsPerBucket);
+  }
+  return visConfig;
+};
```
Comment (fbaligand) [inline public/enh_table_to_ast.ts:39]: Why do you add this for enhanced_table, whereas you don't define it for "document_table" ?

File: public/data_load/kibana_cloned_code/courier.ts [hunk 4]
```diff
@@ -193,14 +193,15 @@ export const handleCourierRequest = async ({
...
+  // DataAdapter is not present in Kibana 7.12. Is this necessary?
+/*   inspectorAdapters.data.setTabularLoader(
     () =>
       buildTabularInspectorData((searchSource as any).tabifiedResponse, {
         queryFilter: filterManager,
         deserializeFieldFormat: getFieldFormats().deserialize,
       }),
     { returnsFormattedValues: true }
-  );
+  ); */
```
Comment (fbaligand) [inline public/data_load/kibana_cloned_code/courier.ts:204]: Well, courier.ts is the mirror of this file:
https://github.com/elastic/kibana/blob/7.12/src/plugins/data/common/search/expressions/esaggs/request_handler.ts

So I invite you to take inspiration from this file.
This should fix "Inspect" feature, where "View: Data" does not work (it is empty).

**Review by fbaligand — CHANGES_REQUESTED** [review]
This is a first review.
Anyway, once again, you did an amazing work!

File: .vscode/launch.json [hunk 5]
```diff
@@ -0,0 +1,20 @@
...
+            "skipFiles": [
+                "<node_internals>/**"
+            ],
+            "program": "${file}",
+            "outFiles": [
+                "${workspaceFolder}/**/*.js"
+            ]
+        }
+    ]
+}
```
[the remark this answers is not in the document]
Comment (cpiment) [reply .vscode/launch.json]: Removed

File: .vscode/settings.json [hunk 6]
```diff
@@ -0,0 +1,2 @@
+{
+}
```
[the remark this answers is not in the document]
Comment (cpiment) [reply .vscode/settings.json]: Removed

File: public/data_load/enhanced-table-request-handler.js [hunk 7]
```diff
@@ -32,10 +30,11 @@ export async function enhancedTableRequestHandler ({
   filters,
   inspectorAdapters,
   forceFetch,
-  aggs
+  aggs,
+  queryFilter
 }) {
 
-  const { filterManager } = getQueryService();
+  const filterManager = queryFilter
```
[the remark this answers is not in the document]
Comment (cpiment) [reply public/data_load/enhanced-table-request-handler.js:37]:
```suggestion
  const filterManager = queryFilter;
```

File: public/data_load/enhanced-table-request-handler.js (hunk 2 above)
[the remark this answers is not in the document]
Comment (cpiment) [reply public/data_load/enhanced-table-request-handler.js:123]:
```suggestion
  response.columns = await enrichColumnsWithAggconfig(response.columns);
```

File: public/data_load/enhanced-table-request-handler.js [hunk 8]
```diff
@@ -148,3 +153,16 @@ export async function enhancedTableRequestHandler ({
   // return elasticsearch response
   return response;
 }
+
+// Adds aggConfig to each column using the search service
+async function enrichColumnsWithAggconfig(columns){
+  const promises = columns.map(async (column) => {
+      column.meta.index = column.meta.sourceParams.indexPatternId
```
[the remark this answers is not in the document]
Comment (cpiment) [reply public/data_load/enhanced-table-request-handler.js:160]:
```suggestion
      column.meta.index = column.meta.sourceParams.indexPatternId;
```

File: public/data_load/enhanced-table-request-handler.js [hunk 9]
```diff
@@ -148,3 +153,16 @@ export async function enhancedTableRequestHandler ({
...
+      const indexPattern = await getSearchService().aggs.datatableUtilities.getIndexPattern(column)
```
[the remark this answers is not in the document]
Comment (cpiment) [reply public/data_load/enhanced-table-request-handler.js:161]:
```suggestion
      const indexPattern = await getSearchService().aggs.datatableUtilities.getIndexPattern(column);
```

File: public/data_load/enhanced-table-request-handler.js [hunk 10]
```diff
@@ -148,3 +153,16 @@ export async function enhancedTableRequestHandler ({
...
+      return {
+          ...column,
+          aggConfig: getSearchService().aggs.createAggConfigs(indexPattern,[column.meta.sourceParams]).aggs[0]
+      }
```
[the remark this answers is not in the document]
Comment (cpiment) [reply public/data_load/enhanced-table-request-handler.js:165]: Done

File: public/data_load/enhanced-table-response-handler.js [hunk 11]
```diff
@@ -98,6 +103,10 @@ function splitTable(columns, rows, $parent) {
   });
 }
 
-export function enhancedTableResponseHandler(response) {
+export async function enhancedTableResponseHandler(response, dimensions) {
+  console.log(dimensions);
```
[the remark this answers is not in the document]
Comment (cpiment) [reply public/data_load/enhanced-table-response-handler.js]: Removed

File: public/doc_table_fn.ts [hunk 12]
```diff
@@ -0,0 +1,149 @@
...
+
+import { get } from 'lodash';
+import { i18n } from '@kbn/i18n';
+import { PersistedState } from '../../../src/plugins/visualizations/public';
+import { ExpressionFunctionDefinition, Render } from '../../../src/plugins/expressions/public';
+import { getIndexPatterns, getFilterManager, getSearchService, getVisualization } from './services';
+import { enhancedTableRequestHandler } from './data_load/enhanced-table-request-handler';
+import { documentTableResponseHandler } from './data_load/document-table-response-handler';
+
+import { DOC_TABLE_VIS_NAME } from './types'
```
[the remark this answers is not in the document]
Comment (cpiment) [reply public/doc_table_fn.ts]: Removed, now this file is `visualization_fn.ts`

File: public/enh_table_to_ast.ts (hunk 3 above)
Comment (cpiment) [reply public/enh_table_to_ast.ts]: This code was copied from the Data Table visualization `toExpressionAst` from Kibana 7.12. I made this at the beginning of my work and I did not really understood what I was copying. I have removed this code and unified both `enh_table_to_ast.ts`and `doc_table_to_ast.ts` into `to_ast.ts` since now both functions are almost identical

File: public/data_load/enhanced-table-request-handler.js (hunk 1 above)
Comment (cpiment) [reply public/data_load/enhanced-table-request-handler.js:91]: After updating `courier.ts` and `enhanced-table-request-handler.js` to be as similar as possible to `esaggs`, Data tab of the inspect window was still empty.

Looking at other visualizations in Kibana I found some lines inside the expression functions:
```js
    if (inspectorAdapters?.tables) {
      inspectorAdapters.tables.logDatatable('default', input.tables[0]);
    }
```
It seems that `logDataTable`is the responsible to pass that data to the Inspect window. I'm using `input.tables[0]`. Can that `tables` array have more than one entry? If that is the case maybe `logDataTable`should be called for every table of the array. What do you think?

File: public/data_load/enhanced-table-request-handler.js (hunk 2 above)
Comment (cpiment) [reply public/data_load/enhanced-table-request-handler.js]: Modified with your suggestion. Thanks!

File: public/data_load/kibana_cloned_code/courier.ts (hunk 4 above)
Comment (cpiment) [reply public/data_load/kibana_cloned_code/courier.ts:214]: Please, see [this comment](https://github.com/fbaligand/kibana-enhanced-table/pull/230#discussion_r748653607)

File: public/data_load/enhanced-table-request-handler.js (hunk 1 above)
Comment (cpiment) [reply public/data_load/enhanced-table-request-handler.js:91]: I made some more text and I have fixed it passing to `logDataTable` the result of the requestHandler instead of the result of the responseHandler. I think that will work event if there is more than one table after the responseHandler runs. Please see this [commit](https://github.com/cpiment/kibana-enhanced-table/commit/df09017ca51648335abb0b0c29602ef8f213dd13)

**Comment by cpiment** [comment]
Hi @fbaligand ,

I made this change (f1ca687175f8b78190793f17a4f9d719314e329d) that enables the scroll bars in the panel when the visualization doesn't fit. I solved it by adding the `visualization` class to the Angular template of the visualizations, because I saw that it was done in every other visualization in Kibana.

I don't know if adding that class can cause some side effect that I'm not seeing right now. What do you think?

**Comment by fbaligand** [comment]
All the problem is that mix AngularJS and React is not so easy :)

But you can see that VisualizationContainer is used in classic "Data Table" here:
https://github.com/elastic/kibana/blob/7.12/src/plugins/vis_type_table/public/table_vis_renderer.tsx

And the equivalent file in enhanced table is `vis_legacy_renderer.tsx`, line 54.
But the problem starts there. `vis_legacy_renderer.tsx` does a complicated work to render a AngularJS plugin into a React architecture.
I think that the idea is to "inject" VisualizationContainer here. But how? Good question... I have not the answer.
You can eventually try but I guess that it is not so easy.

File: public/data_load/enhanced-table-request-handler.js (hunk 1 above)
Comment (fbaligand) [reply public/data_load/enhanced-table-request-handler.js:91]: Nice!
I just tested it, it works well now!

**Comment by fbaligand** [comment]
Thanks for all the new fixes!

I just detected a bug: the buttons "Filter for value" / "Filter out value" do not work :(
It raises this error:
image

**Comment by fbaligand** [comment]
Well, one global remark: for each Kibana minor version, I compare kibana/src/plugins/vis_type_table in target branch (here 7.12) with previous branch (here 7.11), and then I try to apply each change that is relevant.
Especially, all glue code to make work AngularJS into Kibana new platform.
So I invite to do so.

Especially, look at `vis_controller.ts`, you will see several interesting things, notably for 'visualization' class, but also for the "filtering" bug.

**Comment by cpiment** [comment]
Hi @fbaligand 

I added some lines that were missing in `vis_controller.ts` comparing it with the Data Table. I have seen that `rows.js` is very different in Data Table and Enhanced Table, and I don't really know how to merge that differences.

Regarding the filter/filter out issue, I have added the `$scope.filter` in the renderer, but that `$scope` is different from the one that uses `createFilterableCell` in `rows.js` so the error still persists because that filter function is not initialized anywhere. I have tried to find where in Data Table plugin it does get initialized but I have not been able to find it... any ideas?

Thanks for all your help with this

**Comment by fbaligand** [comment]
I saw that in the meantime, you finally found what to change to fix filtering bug => that's great!

File: public/data_load/kibana_cloned_code/courier.ts [hunk 13]
```diff
@@ -116,9 +121,10 @@ export const handleCourierRequest = async ({
   // since the last request
   const shouldQuery = forceFetch || (searchSource as any).lastQuery !== queryHash;
 
+  let request;
   if (shouldQuery) {
     inspectorAdapters.requests.reset();
-    const request = inspectorAdapters.requests.start(
```
Comment (fbaligand) [inline public/data_load/kibana_cloned_code/courier.ts:121]: Why this change?
you use  `request` only in this `if` block, and you instanciate `request` variable only one time.
So I think that it is better to keep this code as before.

File: public/data_load/kibana_cloned_code/courier.ts [hunk 14]
```diff
@@ -186,21 +197,21 @@ export const handleCourierRequest = async ({
...
+  // DataAdapter is not present in Kibana 7.12. Is this necessary?
+/*   inspectorAdapters.data.setTabularLoader(
     () =>
       buildTabularInspectorData((searchSource as any).tabifiedResponse, {
         queryFilter: filterManager,
         deserializeFieldFormat: getFieldFormats().deserialize,
       }),
     { returnsFormattedValues: true }
-  );
+  ); */
```
Comment (fbaligand) [inline public/data_load/kibana_cloned_code/courier.ts:214]: After numerous functional tests, it works great, especially "Inspect" feature.
So you can definitively remove this commented block.

File: public/visualization_fn.ts [hunk 15]
```diff
@@ -0,0 +1,168 @@
...
+ */
+
+import { get } from 'lodash';
+import { i18n } from '@kbn/i18n';
+import { PersistedState } from '../../../src/plugins/visualizations/public';
+import { ExpressionFunctionDefinition, Render } from '../../../src/plugins/expressions/public';
+import { getIndexPatterns, getFilterManager, getSearchService, getVisualization } from './services';
+import { enhancedTableRequestHandler } from './data_load/enhanced-table-request-handler';
+import { enhancedTableResponseHandler } from './data_load/enhanced-table-response-handler';
+import { documentTableResponseHandler } from './data_load/document-table-response-handler';
```
Comment (fbaligand) [inline public/visualization_fn.ts:16]: could you merge these 2 lines in only one line?

File: public/visualization_fn.ts [hunk 16]
```diff
@@ -0,0 +1,168 @@
...
+
+import { DOC_TABLE_VIS_NAME } from './types';
+import { ENH_TABLE_VIS_NAME }  from './types';
+import { VisName } from './types';
```
Comment (fbaligand) [inline public/visualization_fn.ts:20]: could you merge these 3 lines in only one line?

File: public/visualization_fn.ts [hunk 17]
```diff
@@ -0,0 +1,168 @@
+/*
```
Comment (fbaligand) [inline public/visualization_fn.ts:1]: could you move this file to `data_load` sub folder?

**Review by fbaligand — CHANGES_REQUESTED** [review]
This is the final review.
Globally, it is fine. I did a full functional review and detected no bug, so that's great!
I just requested few last changes and then, it will be OK to merge!

File: public/data_load/enhanced-table-request-handler.js [hunk 18]
```diff
@@ -85,22 +85,25 @@ export async function enhancedTableRequestHandler ({
     });
   }
 
-  inspectorAdapters.requests = new RequestAdapter();
-  inspectorAdapters.data = new DataAdapter();
+  //inspectorAdapters.requests is already initialized
+  //inspectorAdapters.requests = new RequestAdapter();
+  // DataAdapter is not present in Kibana 7.12. Is this necessary?
+  //inspectorAdapters.data = new DataAdapter();
```
[the remark this answers is not in the document]
Comment (cpiment) [reply public/data_load/enhanced-table-request-handler.js]: I have removed the lines.

File: public/data_load/enhanced-table-request-handler.js [hunk 19]
```diff
@@ -113,9 +116,18 @@ export async function enhancedTableRequestHandler ({
...
-  });
+  }); */
+
+  //Add aggConfig to each column
+  response.columns = response.columns.map( ( column ) => {
+    return {
+      ...column,
+      aggConfig: aggs.byId(column.meta.sourceParams.id),
+    }
+  })
```
[the remark this answers is not in the document]
Comment (cpiment) [reply public/data_load/enhanced-table-request-handler.js]: Changed the code as requested. Much cleaner this way!

File: package.json [hunk 20]
```diff
@@ -12,10 +12,12 @@
     "es": "node ../../scripts/es",
     "lint": "../../node_modules/.bin/eslint .",
     "start": "cd ../.. && node scripts/kibana --dev --oss",
-    "test:server": "../../node_modules/.bin/plugin-helpers test:server",
-    "test:browser": "../../node_modules/.bin/plugin-helpers test:browser",
-    "build": "../../node_modules/.bin/plugin-helpers build",
-    "compile": "rm -rf ./target && node ../../scripts/build_kibana_platform_plugins.js --scan-dir . --filter enhancedTable"
+    "start-full": "cd ../.. && node scripts/kibana --dev",
```
[the remark this answers is not in the document]
Comment (cpiment) [reply package.json]: Changed the task name to "start-xpack"

File: public/data_load/kibana_cloned_code/courier.ts (hunk 13 above)
Comment (cpiment) [reply public/data_load/kibana_cloned_code/courier.ts]: I made this change while updating the courier.ts file from the esaggs module in Kibana 7.12.1 while searching what was the problem with the data inspector not working. In the esaggs module is the `request`variable is instantiated this way and that's why I made the change.

I have reverted the change as requested

File: public/data_load/kibana_cloned_code/courier.ts (hunk 14 above)
Comment (cpiment) [reply public/data_load/kibana_cloned_code/courier.ts]: I have removed the whole block as requested

File: public/visualization_fn.ts (hunk 15 above)
Comment (cpiment) [reply public/visualization_fn.ts]: Those two functions come from two different files, so I don't think I can merge them in just one line. Or maybe there is a way to do that merge that I don't know... I'm quite unexperienced with typescript :P

File: public/visualization_fn.ts (hunk 16 above)
Comment (cpiment) [reply public/visualization_fn.ts]: Merged the three imports in one line

File: public/visualization_fn.ts (hunk 17 above)
Comment (cpiment) [reply public/visualization_fn.ts]: Moved the file as requested

File: public/visualization_fn.ts (hunk 15 above)
Comment (fbaligand) [reply public/visualization_fn.ts]: Yes, you're right, I misread the two lines :)

**Review by fbaligand — APPROVED** [review]

**Comment by fbaligand** [comment]
Hi @cpiment,

Thanks for these last changes!
This is now all good to me!
=> LGTM!

I will merge this PR soon, and then release enhanced table with Kibana 7.12 suppport!

Thanks again for your awesome work on this PR!

**Comment by cpiment** [comment]
Glad to help!! Thank you for the great plugin. In the next days I will try to start the job for 7.13... Let's hope it is easier than 7.12!

**Comment by fbaligand** [comment]
Great news!
As a first step, you can try to install the plugin as is in kibana 7.13 (just change kibana version in kibana.json). 
With luck, it can work as is!
Actually, the same code works for kibana 7.9, 7.10 and 7.11.

**Comment by fbaligand** [comment]
Hi @cpiment,

I just pushed several changes on master branch (essentially eslint fixes).
So if you aim to add support for Kibana 7.13, I invite you to do first a "git pull" :)

**Comment by fbaligand** [comment]
Hi @cpiment,

As you probably already saw, I published enhanced-table plugin v1.12.0, with Kibana 7.12 support:
https://github.com/fbaligand/kibana-enhanced-table/releases/tag/v1.12.0

You can notice that I mention you :)
